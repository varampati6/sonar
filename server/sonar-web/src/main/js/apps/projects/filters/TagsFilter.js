/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
//@flow
import React from 'react';
import { debounce, difference, sortBy, size } from 'lodash';
import Filter from './Filter';
import FilterHeader from './FilterHeader';
import SearchableFilterFooter from './SearchableFilterFooter';
import SearchableFilterOption from './SearchableFilterOption';
import { searchProjectTags } from '../../../api/components';

type Props = {
  query: {},
  router: { push: ({ pathname: string, query?: {} }) => void },
  value?: Array<string>,
  facet?: {},
  isFavorite?: boolean,
  organization?: {},
  maxFacetValue?: number
};

type State = {
  isLoading: boolean,
  search: string,
  tags: Array<string>
};

const LIST_SIZE = 10;

export default class TagsFilter extends React.PureComponent {
  getSearchOptions: () => [{ label: string, value: string }];
  props: Props;
  state: State;
  property: string;

  constructor(props: Props) {
    super(props);
    this.state = {
      isLoading: false,
      search: '',
      tags: []
    };
    this.property = 'tags';
    this.handleSearch = debounce(this.handleSearch.bind(this), 250);
  }

  getSearchOptions(facet: {}, tags: Array<string>) {
    let tagsCopy = [...tags];
    if (facet) {
      tagsCopy = difference(tagsCopy, Object.keys(facet));
    }
    return tagsCopy.slice(0, LIST_SIZE).map(tag => ({ label: tag, value: tag }));
  }

  handleSearch = (search?: string) => {
    if (search !== this.state.search) {
      search = search || '';
      this.setState({ search, isLoading: true });
      searchProjectTags({
        q: search,
        ps: size(this.props.facet || {}) + LIST_SIZE
      }).then(result => {
        this.setState({ isLoading: false, tags: result.tags });
      });
    }
  };

  getSortedOptions(facet: {} = {}) {
    return sortBy(Object.keys(facet), [option => -facet[option], option => option]);
  }

  getFacetValueForOption = (facet: {}, option: string) => facet[option];

  renderOption = (option: string) => <SearchableFilterOption optionKey={option} />;

  render() {
    return (
      <Filter
        property={this.property}
        options={this.getSortedOptions(this.props.facet)}
        query={this.props.query}
        renderOption={this.renderOption}
        value={this.props.value}
        facet={this.props.facet}
        maxFacetValue={this.props.maxFacetValue}
        isFavorite={this.props.isFavorite}
        organization={this.props.organization}
        getFacetValueForOption={this.getFacetValueForOption}
        highlightUnder={1}
        header={<FilterHeader name="Tags" />}
        footer={
          <SearchableFilterFooter
            property={this.property}
            query={this.props.query}
            options={this.getSearchOptions(this.props.facet, this.state.tags)}
            isLoading={this.state.isLoading}
            onOpen={this.handleSearch}
            onInputChange={this.handleSearch}
            isFavorite={this.props.isFavorite}
            organization={this.props.organization}
            router={this.props.router}
          />
        }
      />
    );
  }
}
