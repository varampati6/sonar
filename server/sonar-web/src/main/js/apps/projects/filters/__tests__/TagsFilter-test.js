/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import TagsFilter from '../TagsFilter';

const tags = ['lang', 'sonar', 'csharp', 'dotnet', 'it', 'net'];
const tagsFacet = { lang: 4, sonar: 3, csharp: 1 };
const fakeRouter = { push: () => {} };

it('should render the tags without the ones in the facet', () => {
  const wrapper = shallow(
    <TagsFilter query={{ tags: null }} router={fakeRouter} facet={tagsFacet} />
  );
  expect(wrapper).toMatchSnapshot();
  wrapper.setState({ tags });
  expect(wrapper).toMatchSnapshot();
});

it('should render the tags facet with the selected tags', () => {
  const wrapper = shallow(
    <TagsFilter
      query={{ tags: ['lang', 'sonar'] }}
      value={['lang', 'sonar']}
      router={fakeRouter}
      facet={tagsFacet}
      isFavorite={true}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('Filter').shallow()).toMatchSnapshot();
});

it('should render maximum 10 tags in the searchbox results', () => {
  const wrapper = shallow(
    <TagsFilter
      query={{ languages: ['java', 'ad'] }}
      value={['java', 'ad']}
      router={fakeRouter}
      facet={{ ...tagsFacet, ad: 1 }}
      isFavorite={true}
    />
  );
  wrapper.setState({ tags: [...tags, 'aa', 'ab', 'ac', 'ad', 'ae', 'af', 'ag', 'ah', 'ai'] });
  expect(wrapper).toMatchSnapshot();
});
