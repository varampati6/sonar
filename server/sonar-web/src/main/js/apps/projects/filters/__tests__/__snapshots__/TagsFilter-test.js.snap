// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render maximum 10 tags in the searchbox results 1`] = `
<Filter
  facet={
    Object {
      "ad": 1,
      "csharp": 1,
      "lang": 4,
      "sonar": 3,
    }
  }
  footer={
    <SearchableFilterFooter
      isFavorite={true}
      isLoading={false}
      onInputChange={[Function]}
      onOpen={[Function]}
      options={
        Array [
          Object {
            "label": "dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "it",
            "value": "it",
          },
          Object {
            "label": "net",
            "value": "net",
          },
          Object {
            "label": "aa",
            "value": "aa",
          },
          Object {
            "label": "ab",
            "value": "ab",
          },
          Object {
            "label": "ac",
            "value": "ac",
          },
          Object {
            "label": "ae",
            "value": "ae",
          },
          Object {
            "label": "af",
            "value": "af",
          },
          Object {
            "label": "ag",
            "value": "ag",
          },
          Object {
            "label": "ah",
            "value": "ah",
          },
        ]
      }
      property="tags"
      query={
        Object {
          "languages": Array [
            "java",
            "ad",
          ],
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Tags"
    />
  }
  highlightUnder={1}
  isFavorite={true}
  options={
    Array [
      "lang",
      "sonar",
      "ad",
      "csharp",
    ]
  }
  property="tags"
  query={
    Object {
      "languages": Array [
        "java",
        "ad",
      ],
    }
  }
  renderOption={[Function]}
  value={
    Array [
      "java",
      "ad",
    ]
  }
/>
`;

exports[`should render the tags facet with the selected tags 1`] = `
<Filter
  facet={
    Object {
      "csharp": 1,
      "lang": 4,
      "sonar": 3,
    }
  }
  footer={
    <SearchableFilterFooter
      isFavorite={true}
      isLoading={false}
      onInputChange={[Function]}
      onOpen={[Function]}
      options={Array []}
      property="tags"
      query={
        Object {
          "tags": Array [
            "lang",
            "sonar",
          ],
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Tags"
    />
  }
  highlightUnder={1}
  isFavorite={true}
  options={
    Array [
      "lang",
      "sonar",
      "csharp",
    ]
  }
  property="tags"
  query={
    Object {
      "tags": Array [
        "lang",
        "sonar",
      ],
    }
  }
  renderOption={[Function]}
  value={
    Array [
      "lang",
      "sonar",
    ]
  }
/>
`;

exports[`should render the tags facet with the selected tags 2`] = `
<div
  className="search-navigator-facet-box"
  data-key="tags"
>
  <FilterHeader
    name="Tags"
  />
  <div
    className="search-navigator-facet-list"
  >
    <Link
      className="facet search-navigator-facet projects-facet active"
      data-key="lang"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/favorite",
          "query": Object {
            "tags": "sonar",
          },
        }
      }
    >
      <span
        className="facet-name"
      >
        <SearchableFilterOption
          optionKey="lang"
        />
      </span>
      <span
        className="facet-stat"
      >
        4
      </span>
    </Link>
    <Link
      className="facet search-navigator-facet projects-facet active"
      data-key="sonar"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/favorite",
          "query": Object {
            "tags": "lang",
          },
        }
      }
    >
      <span
        className="facet-name"
      >
        <SearchableFilterOption
          optionKey="sonar"
        />
      </span>
      <span
        className="facet-stat"
      >
        3
      </span>
    </Link>
    <Link
      className="facet search-navigator-facet projects-facet"
      data-key="csharp"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/favorite",
          "query": Object {
            "tags": "lang,sonar,csharp",
          },
        }
      }
    >
      <span
        className="facet-name"
      >
        <SearchableFilterOption
          optionKey="csharp"
        />
      </span>
      <span
        className="facet-stat"
      >
        1
      </span>
    </Link>
  </div>
  <SearchableFilterFooter
    isFavorite={true}
    isLoading={false}
    onInputChange={[Function]}
    onOpen={[Function]}
    options={Array []}
    property="tags"
    query={
      Object {
        "tags": Array [
          "lang",
          "sonar",
        ],
      }
    }
    router={
      Object {
        "push": [Function],
      }
    }
  />
</div>
`;

exports[`should render the tags without the ones in the facet 1`] = `
<Filter
  facet={
    Object {
      "csharp": 1,
      "lang": 4,
      "sonar": 3,
    }
  }
  footer={
    <SearchableFilterFooter
      isLoading={false}
      onInputChange={[Function]}
      onOpen={[Function]}
      options={Array []}
      property="tags"
      query={
        Object {
          "tags": null,
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Tags"
    />
  }
  highlightUnder={1}
  options={
    Array [
      "lang",
      "sonar",
      "csharp",
    ]
  }
  property="tags"
  query={
    Object {
      "tags": null,
    }
  }
  renderOption={[Function]}
/>
`;

exports[`should render the tags without the ones in the facet 2`] = `
<Filter
  facet={
    Object {
      "csharp": 1,
      "lang": 4,
      "sonar": 3,
    }
  }
  footer={
    <SearchableFilterFooter
      isLoading={false}
      onInputChange={[Function]}
      onOpen={[Function]}
      options={
        Array [
          Object {
            "label": "dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "it",
            "value": "it",
          },
          Object {
            "label": "net",
            "value": "net",
          },
        ]
      }
      property="tags"
      query={
        Object {
          "tags": null,
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Tags"
    />
  }
  highlightUnder={1}
  options={
    Array [
      "lang",
      "sonar",
      "csharp",
    ]
  }
  property="tags"
  query={
    Object {
      "tags": null,
    }
  }
  renderOption={[Function]}
/>
`;
